src/staticdata.c

```diff
@@ -32,61 +32,174 @@ extern "C" {
 
 // An array of references that need to be restored from the sysimg
 // This is a manually constructed dual of the gvars array, which would be produced by codegen for Julia code, for C.
-static void *const _tags[] = {
-         // builtin types
-         &jl_any_type, &jl_symbol_type, &jl_ssavalue_type, &jl_datatype_type, &jl_slotnumber_type,
-         &jl_simplevector_type, &jl_array_type, &jl_typedslot_type,
-         &jl_expr_type, &jl_globalref_type, &jl_string_type,
-         &jl_module_type, &jl_tvar_type, &jl_method_instance_type, &jl_method_type, &jl_code_instance_type,
-         &jl_linenumbernode_type, &jl_lineinfonode_type,
-         &jl_gotonode_type, &jl_quotenode_type, &jl_gotoifnot_type, &jl_argument_type, &jl_returnnode_type,
-         &jl_const_type, &jl_partial_struct_type, &jl_method_match_type,
-         &jl_pinode_type, &jl_phinode_type, &jl_phicnode_type, &jl_upsilonnode_type,
-         &jl_type_type, &jl_bottom_type, &jl_ref_type, &jl_pointer_type, &jl_llvmpointer_type,
-         &jl_vararg_type, &jl_abstractarray_type,
-         &jl_densearray_type, &jl_nothing_type, &jl_function_type, &jl_typeofbottom_type,
-         &jl_unionall_type, &jl_typename_type, &jl_builtin_type, &jl_code_info_type,
-         &jl_task_type, &jl_uniontype_type, &jl_abstractstring_type,
-         &jl_array_any_type, &jl_intrinsic_type, &jl_abstractslot_type,
-         &jl_methtable_type, &jl_typemap_level_type, &jl_typemap_entry_type,
-         &jl_voidpointer_type, &jl_uint8pointer_type, &jl_newvarnode_type,
-         &jl_anytuple_type_type, &jl_anytuple_type, &jl_namedtuple_type, &jl_emptytuple_type,
-         &jl_array_symbol_type, &jl_array_uint8_type, &jl_array_int32_type,
-         &jl_int32_type, &jl_int64_type, &jl_bool_type, &jl_uint8_type,
-         &jl_uint32_type, &jl_uint64_type, &jl_char_type, &jl_weakref_type,
-         &jl_int8_type, &jl_int16_type, &jl_uint16_type,
-         &jl_float16_type, &jl_float32_type, &jl_float64_type, &jl_floatingpoint_type,
-         &jl_number_type, &jl_signed_type,
-         // special typenames
-         &jl_tuple_typename, &jl_pointer_typename, &jl_llvmpointer_typename, &jl_array_typename, &jl_type_typename,
-         &jl_vararg_typename, &jl_namedtuple_typename,
-         &jl_vecelement_typename,
-         // special exceptions
-         &jl_errorexception_type, &jl_argumenterror_type, &jl_typeerror_type,
-         &jl_methoderror_type, &jl_loaderror_type, &jl_initerror_type,
-         &jl_undefvarerror_type, &jl_stackovf_exception, &jl_diverror_exception,
-         &jl_interrupt_exception, &jl_boundserror_type, &jl_memory_exception,
-         &jl_undefref_exception, &jl_readonlymemory_exception,
+jl_value_t **const*const get_tags(void) {
+    // Make sure to keep an extra slot at the end to sentinel length
+    static void * _tags[146] = {NULL};
+
+    // Lazyily-initialize this list
+    if (_tags[0] == NULL) {
+        // builtin types
+        _tags[  0] = &jl_any_type;
```

Comment:
Just a drive-by comment — something like the following could make this a little neater, I guess?

```C
    const int ntags = 146;
    static void * _tags[ntags] = {NULL};
    if (_tags[0] == NULL) {
        int i = 0;
#define GVAR(sym) = _tags[i++] = &sym;
        GVAR(jl_any_type); GVAR(jl_symbol_type); ...
        // all globals ...

#undef GVAR
        assert(i == ntags);
```

Reply:
I liked this so much, I blindly used it, and got a little confused by the error messages until I saw that you have been writing a little too much Julia code.  ;)

```
#define GVAR(sym) = _tags[i++] = &(sym)
```

Reply:
Haha, I had to stare at this for a full 30 s until I saw that extra `=` sign.  You're right, too much Julia lately and no enough C :laughing: